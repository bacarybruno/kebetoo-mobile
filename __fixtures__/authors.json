[{
  "uid": "1",
  "email": "bacarybruno@kebetoo.com",
  "displayName": "Bruno Bodian",
  "photoURL": "https://avatars1.githubusercontent.com/u/14147533"
}, {
  "uid": "2",
  "email": "johndoe@kebetoo.com",
  "displayName": "John Doe",
  "photoURL": null
}]