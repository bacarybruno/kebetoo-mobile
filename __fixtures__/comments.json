[
  {
    "_id": "5eedd9d43acba504d4a1cfbf",
    "author": "1",
    "content": "Fgyjj",
    "createdAt": "2020-06-20T09:41:40.401Z",
    "updatedAt": "2020-06-20T09:41:40.405Z",
    "__v": 0,
    "post": {
      "_id": "5eedd9a43acba504d4a1cfbb",
      "content": "Fhgjjyjj",
      "author": "1",
      "createdAt": "2020-06-20T09:40:52.790Z",
      "updatedAt": "2020-06-20T09:40:52.841Z",
      "__v": 0,
      "audio": {
        "_id": "5eedd9a43acba504d4a1cfbc",
        "name": "PTT-20200620-6",
        "alternativeText": null,
        "caption": null,
        "hash": "PTT_20200620_6_5e895bdeae",
        "ext": ".mp3",
        "mime": "audio/mp3",
        "size": 13.24,
        "url": "/uploads/PTT_20200620_6_5e895bdeae.mp3",
        "provider": "local",
        "width": null,
        "height": null,
        "related": ["5eedd9a43acba504d4a1cfbb"],
        "createdAt": "2020-06-20T09:40:52.832Z",
        "updatedAt": "2020-06-20T09:40:52.839Z",
        "__v": 0,
        "id": "5eedd9a43acba504d4a1cfbc"
      },
      "id": "5eedd9a43acba504d4a1cfbb"
    },
    "reactions": [
      {
        "type": "love",
        "_id": "5eee1f963acba504d4a1cfd1",
        "author": "1",
        "createdAt": "2020-06-20T14:39:18.794Z",
        "updatedAt": "2020-06-20T14:39:18.807Z",
        "__v": 0,
        "comment": "5eedd9d43acba504d4a1cfbf",
        "id": "5eee1f963acba504d4a1cfd1"
      }
    ],
    "id": "5eedd9d43acba504d4a1cfbf"
  },
  {
    "_id": "5eedd9e83acba504d4a1cfc0",
    "author": "1",
    "createdAt": "2020-06-20T09:42:00.149Z",
    "updatedAt": "2020-06-20T09:42:00.179Z",
    "__v": 0,
    "post": {
      "_id": "5eedd9a43acba504d4a1cfbb",
      "content": "Fhgjjyjj",
      "author": "1",
      "createdAt": "2020-06-20T09:40:52.790Z",
      "updatedAt": "2020-06-20T09:40:52.841Z",
      "__v": 0,
      "audio": {
        "_id": "5eedd9a43acba504d4a1cfbc",
        "name": "PTT-20200620-6",
        "alternativeText": null,
        "caption": null,
        "hash": "PTT_20200620_6_5e895bdeae",
        "ext": ".mp3",
        "mime": "audio/mp3",
        "size": 13.24,
        "url": "/uploads/PTT_20200620_6_5e895bdeae.mp3",
        "provider": "local",
        "width": null,
        "height": null,
        "related": ["5eedd9a43acba504d4a1cfbb"],
        "createdAt": "2020-06-20T09:40:52.832Z",
        "updatedAt": "2020-06-20T09:40:52.839Z",
        "__v": 0,
        "id": "5eedd9a43acba504d4a1cfbc"
      },
      "id": "5eedd9a43acba504d4a1cfbb"
    },
    "audio": {
      "_id": "5eedd9e83acba504d4a1cfc1",
      "name": "PTT-20200620-3",
      "alternativeText": null,
      "caption": null,
      "hash": "PTT_20200620_3_1669bbfc5d",
      "ext": ".mp3",
      "mime": "audio/mp3",
      "size": 8.02,
      "url": "/uploads/PTT_20200620_3_1669bbfc5d.mp3",
      "provider": "local",
      "width": null,
      "height": null,
      "related": ["5eedd9e83acba504d4a1cfc0"],
      "createdAt": "2020-06-20T09:42:00.174Z",
      "updatedAt": "2020-06-20T09:42:00.178Z",
      "__v": 0,
      "id": "5eedd9e83acba504d4a1cfc1"
    },
    "reactions": [
      {
        "type": "love",
        "_id": "5eee1fa43acba504d4a1cfd2",
        "author": "1",
        "createdAt": "2020-06-20T14:39:32.870Z",
        "updatedAt": "2020-06-20T14:39:32.878Z",
        "__v": 0,
        "comment": "5eedd9e83acba504d4a1cfc0",
        "id": "5eee1fa43acba504d4a1cfd2"
      }
    ],
    "id": "5eedd9e83acba504d4a1cfc0"
  }
]
