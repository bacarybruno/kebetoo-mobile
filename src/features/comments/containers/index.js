import React, { useCallback, useRef, useMemo } from 'react'
import {
  View, FlatList, KeyboardAvoidingView, Platform,
} from 'react-native'
import { useRoute, useNavigation } from '@react-navigation/native'
import { HeaderBackButton } from '@react-navigation/stack'
import { getBottomSpace } from 'react-native-iphone-x-helper'
import { useSafeAreaInsets } from 'react-native-safe-area-context'

import { Content, Header } from '@app/features/post/containers/basic-post'
import { HeaderBack, NoContent } from '@app/shared/components'
import { strings } from '@app/config'
import {
  useAppColors, useAppStyles, useAudioRecorder, useUser,
} from '@app/shared/hooks'
import { metrics } from '@app/theme'

import createThemedStyles from './styles'
import CommentInput from '../components/comment-input'
import Reactions from '../components/reactions'
import Comment from '../components/comment'
import Swipeable from '../components/swipeable'
import useComments from './hook'

export const NoComments = () => (
  <NoContent title={strings.general.no_content} text={strings.comments.no_content} />
)

export const CommentReply = ({ reply, profile }) => {
  const styles = useAppStyles(createThemedStyles)
  return (
    <View style={styles.replyWrapper}>
      <Comment
        item={reply}
        user={profile.uid}
        authorId={reply.author.id}
        displayName={reply.author.displayName}
        photoURL={reply.author.photoURL}
        avatarSize={30}
      />
    </View>
  )
}

// TODO: paginate comments
// TODO: create custom hook
const Comments = () => {
  const audioRecorder = useAudioRecorder()
  const { params: { post } } = useRoute()
  const navigation = useNavigation()

  const commentInput = useRef()
  const scrollView = useRef()

  const { profile } = useUser()
  const {
    clearToReply,
    loadReplies,
    onComment,
    onCommentContentPress: onCommentPress,
    onSend,
    onSetReply,
    setComment,
    authors,
    comments,
    isLoading,
    toReply,
    replies,
    comment,
  } = useComments(navigation, post, commentInput)
  const insets = useSafeAreaInsets()

  const styles = useAppStyles(createThemedStyles)
  const colors = useAppColors()

  const renderComment = useMemo(() => ({ item }) => {
    if (!item.author) return null

    const checkSelectedComment = (c) => (
      toReply?.id === c.id ? styles.selectedComment : {}
    )

    const onShowReplies = () => loadReplies(item)

    return (
      <View style={styles.comment}>
        <Swipeable
          style={[styles.swipeable, checkSelectedComment(item)]}
          onFulfilled={() => onSetReply(item)}
        >
          <Comment
            item={item}
            user={profile.uid}
            authorId={item.author.id}
            displayName={item.author.displayName}
            photoURL={item.author.photoURL}
            repliesCount={item.replies?.length}
            onShowReplies={onShowReplies}
          />
        </Swipeable>
        {replies[item.id]?.map((reply) => (
          <Swipeable
            onFulfilled={() => onSetReply(reply)}
            style={[styles.swipeableReply, checkSelectedComment(reply)]}
            key={`comment-reply-${reply.id}`}
          >
            <CommentReply reply={reply} profile={profile} />
          </Swipeable>
        ))}
      </View>
    )
  }, [styles, profile, replies, toReply, loadReplies, onSetReply])

  const ListHeaderLeft = useCallback(() => (
    <HeaderBackButton
      onPress={navigation.goBack}
      labelVisible={false}
      backImage={() => (
        <HeaderBack tintColor={colors.textPrimary} />
      )}
    />
  ), [colors.textPrimary, navigation.goBack])

  const ListHeader = useMemo(() => (
    <View style={styles.post}>
      <View style={styles.postHeader}>
        <Header Left={ListHeaderLeft} author={post.author} post={post} size={35} />
        <Content
          post={post}
          style={styles.content}
          mode="comments"
          onPress={onCommentPress}
          originalAuthor={
            post.repost
              ? authors[post.repost.author]
              : post.author
          }
        />
      </View>
      <Reactions post={post} author={profile.uid} comments={comments} onComment={onComment} />
    </View>
  ), [ListHeaderLeft, styles, authors, comments, onComment, onCommentPress, post, profile.uid])

  const keyExtractor = useCallback((item, index) => `comment-${item.id}-${index}`, [])

  return (
    <View style={styles.wrapper}>
      <View style={styles.flexible}>
        <FlatList
          data={comments}
          renderItem={renderComment}
          keyExtractor={keyExtractor}
          ListHeaderComponent={ListHeader}
          ListEmptyComponent={NoComments}
          contentContainerStyle={styles.flatlistContent}
          removeClippedSubviews
          ref={scrollView}
        />
      </View>
      <KeyboardAvoidingView
        style={styles.keyboard}
        enabled={Platform.OS === 'ios'}
        keyboardVerticalOffset={insets.bottom + metrics.marginVertical + getBottomSpace()}
        behavior="padding"
      >
        <CommentInput
          onChange={setComment}
          onSend={onSend}
          inputRef={commentInput}
          audioRecorder={audioRecorder}
          value={comment}
          isLoading={isLoading}
          reply={toReply}
          onReplyClose={clearToReply}
        />
      </KeyboardAvoidingView>
    </View>
  )
}

export default React.memo(Comments)
