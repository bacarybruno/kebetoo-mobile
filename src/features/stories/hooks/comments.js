import { useEffect, useCallback, useReducer } from 'react'

import { useUser, usePosts, useAudioRecorder } from '@app/shared/hooks'
import { api } from '@app/shared/services'
import reducer, { initialState, actionTypes } from '@app/features/comments/containers/reducer'

/**
 * Hackernews' hot sort
 * https://medium.com/hacking-and-gonzo/how-hacker-news-ranking-algorithm-works-1d9b0cf2c08d
 */
export const scoreComment = (votes, itemDate, gravity = 1.8) => {
  const hourAge = (Date.now() - itemDate.getTime()) / (1000 * 3600)
  return (votes - 1) / (hourAge + 2 ** gravity)
}

export const sortComments = (comments) => (
  comments
    .map((comment) => ({
      ...comment,
      score: scoreComment(
        comment.reactions.length + comment.replies.length,
        new Date(comment.createdAt),
      ),
    }))
    .sort((a, b) => {
      if (a.score < b.score) return 1
      if (b.score < a.score) return -1
      return 0
    })
)

const mapComments = (comment) => ({
  id: comment.id,
  author: {
    id: comment.author,
    displayName: ' ',
    photoURL: null,
  },
  reactions: [],
})

// TODO: paginate comments
const useComments = (story, commentInput, scrollView) => {
  const audioRecorder = useAudioRecorder()
  const initialComments = story.comments?.map(mapComments)
  if (initialComments) {
    initialState.comments = initialComments
  }

  const [state, dispatch] = useReducer(reducer, initialState)

  const {
    authors, comment, toReply, comments, replies, isLoading,
  } = state

  const { profile } = useUser()

  const { getRepostAuthors } = usePosts()

  useEffect(() => {
    const fetchComments = async () => {
      const result = await api.comments.getByStoryId(story.id)
      dispatch({ type: actionTypes.SET_COMMENTS, payload: sortComments(result) })
    }
    fetchComments()
  }, [story.id])

  useEffect(() => {
    const fetchRepostAuthors = async () => {
      const data = await getRepostAuthors(story)
      dispatch({ type: actionTypes.SET_AUTHORS, payload: data })
    }
    fetchRepostAuthors()
  }, [story, getRepostAuthors])

  const startLoading = useCallback(() => {
    dispatch({ type: actionTypes.START_LOADING })
  }, [])

  const endLoading = useCallback(() => {
    dispatch({ type: actionTypes.END_LOADING })
  }, [])

  const setLoading = useCallback((loading) => {
    if (loading) startLoading()
    else endLoading()
  }, [endLoading, startLoading])

  const onSend = useCallback(async () => {
    setLoading(true)

    try {
      let result = null
      let replyThread = null
      if (toReply) {
        // use the comment reply thread or the base comment
        replyThread = toReply.thread || toReply
      }

      if (audioRecorder.hasRecording) {
        result = await audioRecorder.saveStoryComment(story.id, profile.uid, replyThread)
      } else if (comment.length > 0) {
        result = await api.comments.create({
          author: profile.uid,
          content: comment,
          thread: replyThread ? replyThread.id : null,
          story: replyThread ? null : story.id,
        })
      }
      if (replyThread) {
        dispatch({
          type: actionTypes.ADD_REPLIES,
          payload: { threadId: replyThread.id, replies: result },
        })
      } else {
        dispatch({ type: actionTypes.ADD_COMMENT, payload: result })
        setTimeout(() => scrollView?.current?.scrollToEnd(), 200)
      }
    } finally {
      // cleanup
      commentInput.current?.clear()
      dispatch({ type: actionTypes.CLEAR_TO_REPLY })
      dispatch({ type: actionTypes.CLEAR_COMMENT })
      setLoading(false)
    }
  }, [setLoading, toReply, audioRecorder, comment, story.id, profile.uid, commentInput, scrollView])

  const onSetReply = useCallback((item) => {
    commentInput.current?.focus()
    dispatch({ type: actionTypes.SET_TO_REPLY, payload: item })
  }, [commentInput])

  const loadReplies = useCallback(async (thread) => {
    if (!replies[thread.id]) {
      dispatch({
        type: actionTypes.SET_REPLIES,
        payload: { threadId: thread.id, replies: thread.replies.map(mapComments) },
      })
    }
    const loadedReplies = await api.comments.getReplies(thread.id)
    dispatch({
      type: actionTypes.SET_REPLIES,
      payload: { threadId: thread.id, replies: loadedReplies },
    })
  }, [replies])

  const onComment = useCallback(() => commentInput.current?.focus(), [])

  const clearToReply = useCallback(() => {
    dispatch({ type: actionTypes.CLEAR_TO_REPLY })
  }, [])

  const setComment = useCallback((payload) => {
    dispatch({ type: actionTypes.SET_COMMENT, payload })
  }, [])

  return {
    onSend,
    onComment,
    onSetReply,
    setComment,
    loadReplies,
    clearToReply,
    audioRecorder,
    isLoading,
    authors,
    comments,
    toReply,
    replies,
    comment,
  }
}

export default useComments
