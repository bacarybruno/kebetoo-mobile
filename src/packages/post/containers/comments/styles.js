import { StyleSheet } from 'react-native'

import colors from 'Kebetoo/src/theme/colors'
import metrics from 'Kebetoo/src/theme/metrics'

export const borderRadiusSize = 55
export const paddingHorizontal = metrics.marginHorizontal * 1.5
export const bottomBarSize = 45

export default StyleSheet.create({
  wrapper: {
    flex: 1,
  },
  flexible: {
    flex: 1,
  },
  reactionsContainer: {
    height: borderRadiusSize,
    borderTopLeftRadius: borderRadiusSize,
    borderTopRightRadius: borderRadiusSize,
    paddingHorizontal,
    backgroundColor: colors.background,
  },
  reactions: {
    flex: 1,
    justifyContent: 'flex-end',
  },
  flatlistContent: {
    flexGrow: 1,
    backgroundColor: colors.background,
  },
  post: {
    width: '100%',
    minHeight: metrics.screenHeight * 0.2,
    backgroundColor: colors.input,
    paddingTop: metrics.marginHorizontal,
    marginBottom: metrics.marginHorizontal,
  },
  content: {
    marginBottom: 0,
    borderRadius: 0,
    backgroundColor: 'transparent',
    padding: metrics.marginHorizontal,
    paddingTop: 0,
  },
  commentInputWrapper: {
    backgroundColor: colors.background,
    paddingHorizontal: metrics.marginHorizontal,
    width: '100%',
    flexDirection: 'row',
    alignItems: 'center',
    paddingBottom: 10,
  },
  textInputSize: {
    height: bottomBarSize,
    minHeight: bottomBarSize,
  },
  textInputWrapper: {
    borderRadius: 100,
  },
  draggableContainer: {
    width: '100%',
    alignItems: 'center',
    backgroundColor: 'transparent',
  },
  draggableIcon: {
    width: 100,
    height: 5,
    borderRadius: 5,
    margin: 10,
    backgroundColor: colors.inactive,
  },
  headerBack: {
    width: 24,
    height: 24,
    margin: 3,
    justifyContent: 'center',
    alignItems: 'center',
  },
  comment: {
    marginHorizontal: paddingHorizontal,
    paddingVertical: metrics.marginHorizontal,
    borderTopColor: 'rgba(0, 0, 0, 0.07)',
    borderTopWidth: 1,
    justifyContent: 'center',
    flexDirection: 'row',
  },
  send: {
    width: bottomBarSize,
    height: bottomBarSize,
    borderRadius: bottomBarSize / 2,
    backgroundColor: colors.input,
    marginLeft: 5,
    justifyContent: 'center',
    alignItems: 'center',
    // elevation
    shadowColor: colors.black,
    shadowOffset: {
      width: 0,
      height: 1,
    },
    shadowOpacity: 0.22,
    shadowRadius: 2.22,
    elevation: 3,
  },
  sendIcon: {
    marginLeft: bottomBarSize / 9,
    color: colors.primary,
  },
  backHandler: {
    width: 24,
    height: 24,
    margin: 3,
    justifyContent: 'center',
    alignItems: 'center',
  },
  noContent: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
    alignSelf: 'center',
  },
  loveIcon: {
    paddingLeft: 10,
  },
})
