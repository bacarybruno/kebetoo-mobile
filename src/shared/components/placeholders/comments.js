import React from 'react'
import {
  Placeholder, PlaceholderMedia, PlaceholderLine, Fade,
} from 'rn-placeholder'

const PlaceholderIcon = () => (
  <PlaceholderMedia size={15} style={{ marginLeft: 10 }} />
)

const PlaceholderAvatar = () => (
  <PlaceholderMedia isRound style={{ marginRight: 10 }} />
)

const CommentPlaceholder = () => (
  <Placeholder Animation={Fade} Left={PlaceholderAvatar} Right={PlaceholderIcon}>
    <PlaceholderLine height={8} width={50} />
    <PlaceholderLine height={10} style={{ marginBottom: 5 }} />
    <PlaceholderLine height={10} />
  </Placeholder>
)

export default React.memo(CommentPlaceholder)
