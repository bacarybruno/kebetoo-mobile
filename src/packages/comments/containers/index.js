import React, {
  useState, useEffect, useCallback, useRef, memo, useMemo,
} from 'react'
import { View, FlatList } from 'react-native'
import { useRoute, useNavigation } from '@react-navigation/native'
import { HeaderBackButton } from '@react-navigation/stack'
import { useSelector, useDispatch } from 'react-redux'
import auth from '@react-native-firebase/auth'

import { Content, Header } from 'Kebetoo/src/packages/post/containers/basic-post'
import HeaderBack from 'Kebetoo/src/shared/components/header-back'
import Text from 'Kebetoo/src/shared/components/text'
import { getUsers } from 'Kebetoo/src/shared/helpers/users'
import { commentsSelector, postsSelector, authorsSelector } from 'Kebetoo/src/redux/selectors'
import * as types from 'Kebetoo/src/redux/types'

import styles from './styles'
import CommentInput from '../components/comment-input'
import Reactions from '../components/reactions'
import Comment from '../components/comment'

export const routeOptions = {}

export const NoContent = () => (
  <View style={styles.noContent}>
    <Text color="primary" size="header" text="No content yet" />
    <Text text="Be the first to add a comment! ðŸ‘‡" />
  </View>
)

// TODO: paginate comments
const Comments = () => {
  const user = auth().currentUser
  const { params } = useRoute()
  const { goBack } = useNavigation()
  const [comments, setComments] = useState([])
  const [authors, setAuthors] = useState({})
  const [comment, setComment] = useState('')
  const commentInput = useRef()
  const normalizedPost = useSelector(postsSelector)[params.id]
  const [post, setPost] = useState((value) => value || normalizedPost)
  const author = useSelector(authorsSelector)[post.author]
  const dispatch = useDispatch()

  const normalizedComments = useSelector(commentsSelector)

  useEffect(() => {
    const coms = Object
      .values(normalizedComments)
      .filter((com) => com.post === params.id)
    setComments(coms)
  }, [normalizedComments, params.id])

  useEffect(() => {
    if (normalizedPost.comments.length !== comments.length) {
      setPost(normalizedPost)
    }
  }, [comments, normalizedPost])

  useEffect(() => {
    const fetchAuthors = async () => {
      if (comments.length > 0) {
        const ids = [...new Set(comments.map((c) => c.author))]
        const newAuthors = ids.filter((id) => !authors[id])

        if (newAuthors.length === 0) return

        const { docs } = await getUsers(newAuthors)
        docs.forEach((doc) => {
          const { displayName: name, photoURL } = doc.data()
          authors[doc.id] = {
            displayName: name,
            photoURL,
          }
        })
        setAuthors({ ...authors })
      }
    }
    fetchAuthors()
  }, [comments, authors])

  const onChangeText = useCallback((value) => {
    setComment(value)
  }, [])

  const onSend = useCallback(async () => {
    if (comment.length > 0) {
      dispatch({
        type: types.COMMENT_POST,
        payload: {
          author: user.uid,
          content: comment,
          post: params.id,
        },
      })
      commentInput.current.clear()
    }
  }, [comment, dispatch, params.id, user.uid])

  const renderComment = useMemo(() => ({ item }) => (
    <View style={styles.comment}>
      <Comment author={authors[item.author]} item={item} />
    </View>
  ), [authors])

  const ListHeaderLeft = useCallback(() => (
    <HeaderBackButton
      onPress={goBack}
      backImage={({ tintColor }) => (
        <HeaderBack tintColor={tintColor} />
      )}
    />
  ), [goBack])

  const onComment = useCallback(() => commentInput.current.focus(), [])

  const ListHeader = useCallback(() => (
    <View style={styles.post}>
      <Header Left={ListHeaderLeft} author={author} post={post} size={35} />
      <Content post={post} style={styles.content} />
      <Reactions
        post={post}
        author={author.id}
        onComment={onComment}
      />
    </View>
  ), [ListHeaderLeft, author, onComment, post])

  const keyExtractor = useCallback((item, index) => `comment-${item.id}-${index}`, [])

  return (
    <View style={styles.wrapper}>
      <View style={styles.flexible}>
        <FlatList
          data={comments}
          renderItem={renderComment}
          keyExtractor={keyExtractor}
          ListHeaderComponent={ListHeader}
          ListEmptyComponent={NoContent}
          contentContainerStyle={styles.flatlistContent}
        />
      </View>
      <CommentInput
        onChange={onChangeText}
        onSend={onSend}
        inputRef={commentInput}
      />
    </View>
  )
}

export default memo(Comments)
