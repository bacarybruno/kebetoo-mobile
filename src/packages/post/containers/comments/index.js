import React, {
  useState, useEffect, useCallback, useRef,
} from 'react'
import { View, FlatList, TouchableOpacity } from 'react-native'
import { useRoute, useNavigation } from '@react-navigation/native'
import { HeaderBackButton } from '@react-navigation/stack'
import { useSelector } from 'react-redux'
import auth from '@react-native-firebase/auth'
import Ionicon from 'react-native-vector-icons/Ionicons'
import moment from 'moment'

import {
  Content, Reactions as BaseReactions, Header,
} from 'Kebetoo/src/packages/post/containers/basic-post'
import Kebeticon from 'Kebetoo/src/shared/icons/kebeticons'
import Text from 'Kebetoo/src/shared/components/text'
import TextInput from 'Kebetoo/src/shared/components/inputs/text'
import Avatar from 'Kebetoo/src/shared/components/avatar'
import colors from 'Kebetoo/src/theme/colors'
import { getUsers } from 'Kebetoo/src/shared/helpers/users'
import { commentPost } from 'Kebetoo/src/shared/helpers/http'

import styles from './styles'

export const routeOptions = {}

export const DraggableIndicator = () => (
  <View style={styles.draggableContainer}>
    <View style={styles.draggableIcon} />
  </View>
)

export const SendButton = ({ onPress }) => (
  <TouchableOpacity style={styles.send} onPress={onPress}>
    <Ionicon style={styles.sendIcon} name="md-send" size={25} color={colors.white} />
  </TouchableOpacity>
)

const CommentInput = ({ onChange, onSend, inputRef }) => (
  <View style={styles.commentInputWrapper}>
    <View style={styles.flexible}>
      <TextInput
        fieldName="comment"
        placeholder="Add a comment"
        onValueChange={onChange}
        ref={inputRef}
        textStyle={styles.textInputSize}
        wrapperStyle={[
          styles.textInputSize,
          styles.textInputWrapper,
        ]}
      />
    </View>
    <SendButton onPress={onSend} />
  </View>
)

const Reactions = ({ post, author }) => (
  <View style={styles.reactionsContainer}>
    <DraggableIndicator />
    <View style={styles.reactions}>
      <BaseReactions post={post} author={author} />
    </View>
  </View>
)

// TODO: paginate comments
const Comments = () => {
  const { params } = useRoute()
  const { goBack } = useNavigation()
  const posts = useSelector((state) => state.postsReducer.posts)
  const post = posts.find((p) => p.id === params.id)
  const author = useSelector((state) => state.postsReducer.authors)[post.author]
  const user = auth().currentUser
  const [authors, setAuthors] = useState({})
  const [comment, setComment] = useState('')
  const commentInput = useRef()

  useEffect(() => {
    const fetchAuthors = async () => {
      if (post.comments.length > 0) {
        const ids = new Set(post.comments.map((c) => c.author))
        const { docs } = await getUsers([...ids])
        const authorsData = {}
        docs.forEach((doc) => {
          const { displayName: name, photoURL } = doc.data()
          authorsData[doc.id] = {
            displayName: name,
            photoURL,
          }
        })
        setAuthors(authorsData)
      }
    }

    fetchAuthors()
  }, [post])

  const onChangeText = useCallback((value) => {
    setComment(value)
  }, [setComment])

  const onSend = useCallback(async () => {
    await commentPost({
      author: user.uid,
      content: comment,
      post: post.id,
    })
    commentInput.current.clear()
  }, [comment, user, post])

  const renderComment = ({ item }) => (
    authors[item.author] ? (
      <View style={styles.comment}>
        <View style={{ flexDirection: 'row', ...styles.flexible }}>
          <View style={{ marginRight: 10 }}>
            <Avatar
              src={authors[item.author].photoURL}
              text={authors[item.author].displayName}
              size={35}
            />
          </View>
          <View style={styles.flexible}>
            <View style={styles.flexible}>
              <View style={{ flexDirection: 'row', ...styles.flexible }}>
                <Text size="sm" text={authors[item.author].displayName} />
                <Text text=" â€¢ " />
                <Text size="sm" text={moment(item.updatedAt).fromNow()} />
              </View>
            </View>
            <Text text={item.content} />
          </View>
        </View>
        <Ionicon name="ios-heart-empty" size={15} />
      </View>
    ) : null
  )

  const ListHeaderLeft = () => (
    <HeaderBackButton
      onPress={goBack}
      backImage={() => (
        <View style={styles.backHandler}>
          <Kebeticon name="chevron-left" size={20} />
        </View>
      )}
    />
  )

  const ListHeader = () => (
    <View style={styles.post}>
      <Header Left={ListHeaderLeft} author={author} post={post} size={35} />
      <Content post={post} style={styles.content} />
      <Reactions post={post} author={author} />
    </View>
  )

  const keyExtractor = (item, index) => `comment-${item.id}-${index}`

  return (
    <View style={styles.wrapper}>
      <FlatList
        data={post.comments}
        renderItem={renderComment}
        keyExtractor={keyExtractor}
        ListHeaderComponent={ListHeader}
        contentContainerStyle={styles.flatlistContent}
      />
      <CommentInput
        onChange={onChangeText}
        onSend={onSend}
        inputRef={commentInput}
      />
    </View>
  )
}

export default Comments
