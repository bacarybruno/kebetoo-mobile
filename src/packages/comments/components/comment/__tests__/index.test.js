import { act } from 'react-test-renderer'

import setupTest from 'Kebetoo/src/config/jest-setup'
import comments from 'Kebetoo/__fixtures__/comments'
import authors from 'Kebetoo/__fixtures__/authors'

import Comment from '../index'

const givenComment = setupTest(Comment)({
  item: {
    ...comments[0],
    reactions: [],
  },
  user: authors[1].uid,
  displayName: authors[0].displayName,
  photoURL: authors[0].photoURL,
})

it('renders Comment', () => {
  const { wrapper } = givenComment()
  expect(wrapper.toJSON()).toMatchSnapshot()
})

it('handles reaction', async () => {
  const { wrapper } = givenComment()
  await act(async () => {
    await wrapper.root.findByProps({ testID: 'reaction-button' }).props.onPress()
  })
  expect(wrapper.root.findByProps({ testID: 'reaction' }).props.name).toBe('md-heart')
  await act(async () => {
    await wrapper.root.findByProps({ testID: 'reaction-button' }).props.onPress()
  })
  expect(wrapper.root.findByProps({ testID: 'reaction' }).props.name).toBe('md-heart-empty')
})
