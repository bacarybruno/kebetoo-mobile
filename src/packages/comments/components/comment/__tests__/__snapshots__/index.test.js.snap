// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders Comment 1`] = `
<View
  accessible={true}
  focusable={true}
  onClick={[Function]}
  onResponderGrant={[Function]}
  onResponderMove={[Function]}
  onResponderRelease={[Function]}
  onResponderTerminate={[Function]}
  onResponderTerminationRequest={[Function]}
  onStartShouldSetResponder={[Function]}
  style={
    Object {
      "flex": 1,
      "flexDirection": "row",
    }
  }
>
  <View
    style={
      Object {
        "marginRight": 10,
      }
    }
  >
    <View
      style={
        Array [
          Object {
            "height": 30,
            "width": 30,
          },
          Object {
            "height": 35,
            "width": 35,
          },
          undefined,
        ]
      }
    >
      <Image
        source={
          Object {
            "uri": "https://avatars1.githubusercontent.com/u/14147533",
          }
        }
        style={
          Array [
            Object {
              "alignItems": "center",
              "backgroundColor": "#008ef4",
              "borderRadius": 15,
              "flex": 1,
              "justifyContent": "center",
            },
            Object {
              "borderRadius": 17.5,
            },
          ]
        }
        testID="image-avatar-https://avatars1.githubusercontent.com/u/14147533"
      />
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <Text
        ellipsizeMode="middle"
        numberOfLines={1}
        style={
          Array [
            Object {
              "fontFamily": "Helvetica",
            },
            Object {
              "color": "rgba(0, 0, 0, 0.8)",
              "fontSize": 14,
              "fontWeight": "normal",
              "opacity": 1,
              "textDecorationLine": "none",
            },
            Object {},
          ]
        }
      >
        
        Bruno Bodian
         â€¢ 
        5 days ago
      </Text>
    </View>
    <Text
      style={
        Array [
          Object {
            "fontFamily": "Helvetica",
          },
          Object {
            "color": "rgba(0, 0, 0, 0.8)",
            "fontSize": 16,
            "fontWeight": "normal",
            "opacity": 1,
            "textDecorationLine": "none",
          },
          Object {},
        ]
      }
    >
      Fgyjj
    </Text>
    <View
      style={
        Object {
          "alignItems": "center",
          "flexDirection": "row",
          "justifyContent": "flex-end",
          "marginTop": 5,
        }
      }
    >
      <View
        accessible={true}
        focusable={true}
        hitSlop={
          Object {
            "bottom": 30,
            "left": 30,
            "right": 30,
            "top": 30,
          }
        }
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "opacity": 1,
          }
        }
        testID="reaction-button"
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "Helvetica",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.8)",
                "fontSize": 12,
                "fontWeight": "bold",
                "opacity": 1,
                "textDecorationLine": "none",
              },
              Object {},
            ]
          }
        >
          0 
        </Text>
        <Text
          allowFontScaling={false}
          style={
            Array [
              Object {
                "fontFamily": "Helvetica",
              },
              Object {
                "color": "rgba(0, 0, 0, 0.8)",
                "fontSize": 16,
                "fontWeight": "normal",
                "opacity": 1,
                "textDecorationLine": "none",
              },
              Object {
                "color": undefined,
                "fontFamily": "Ionicons",
                "fontStyle": "normal",
                "fontWeight": "normal",
              },
            ]
          }
          testID="reaction"
        >
          ionicon-md-heart-empty
        </Text>
      </View>
    </View>
  </View>
</View>
`;
